doc/python/domdec.rst

```diff
@@ -38,6 +40,64 @@ describes the cell groups on the local MPI rank.
         computational cost, hence it may not produce a balanced partition for
         models with cells that have a large variance in computational costs.
 
+.. function:: partition_by_group(recipe, context, groups)
+
+   Construct a :class:`domain_decomposition` that assigns the groups described by
+   the provided list of :class:`group_description` to the local hardware of the calling rank.
+
+   The function expects to be called by each rank in a distributed simulation with the
+   selected groups for that rank. For example, in a simulation of 10 cells on 2 MPI ranks
+   where cells {0, 2, 4, 6, 8} of kind :class:`cable_cell` are expected to be in a single group executed
+   on the GPU on rank 0; and cells {1, 3, 5, 7, 9} of kind :class:`lif_cell` are expected to be in a single
+   group executed on the CPU on rank 1:
+
+   Rank 0 should run:
+
+   .. code-block:: python
+
```

Comment:
Reading these examples of decompositions are actually used, make me wonder whether they should be
moved to the recipe.

Reasoning:
1. Partition-load-balance: needs the recipe anyhow, so could be used like this:
```cxx
struct arb::recipe {
  virtual domain_decomposition decompose(ctx) { return partition_load_balance(...); }
};
```
Concrete recipes might want to override this, but it would remove one non-obvious step from
the construction of a simulation, which -- I think -- our users would appreciate.
2. partition_by_group: needs a lot of the arguments provided by the cell group
  - num_cells and cell_kind by gid for the group makeup
  - recipe and context as before

That said, it's also awkward, since the MPI rank/size is needed and suddenly recipe
becomes dependent on MPI :/ Potentially we could work around that by passing rank/size
to the `decompose` method.

Reply:
I don't think that the `domain_decomposition` should be part of the `recipe`. The `recipe` is used to describe the scientific model: the cells, the connections, the probes etc. The `domain_decomposition` is not related to the model description, it's more technical, it relates the the available hardware resources and to implementation details (how cells are grouped, how expensive they are to simulate). I think it is a good idea to keep these 2 things separate. 
That being said, we can make the `domain_decomposition` immutable after construction, and enforce that the functions that create these objects check their validity after creation.

Reply:
Still, that won't fix the issue that `domain_decomposition` is part of the public interface, so there's nothing stopping the users from using the constructors of the class to create custom `domain_decomposition`s and bypassing the partitioning functions.

Reply:
We could delete all constructors and have the partitioning functions be `friend`s. But I don't like that so much.

Reply:
Yeah, using `friend` would be the way to go, but as ever icky.